connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,266 @@
+    - [Overview](#overview)
+    - [Design](#design)
+        - [Current connection upgrade process](#current-connection-upgrade-process)
+        - [Improved muxer selection](#improved-muxer-selection)
+            - [Muxer selection over TLS](#muxer-selection-over-tls)
+            - [Muxer selection over Noise](#muxer-selection-over-noise)
+                - [Early data specification](#early-data-specification)
+    - [Cross version support](#cross-version-support)
+        - [TLS case](#tls-case)
+        - [Noise case](#noise-case)
+    - [Security](#security)
+    - [Protocol coupling](#protocol-coupling)
+    - [Alternative options considered](#alternative-options-considered)
+
+## Overview
+
+This document discribes an imporvement on the connection upgrade process. The
+goal of the improvement is to reduce the number of RTTs that takes to select the
+muxer of a connection. The solution relies on the ability of the security
+protocol's handshake process to negotiate higher level protocols, which enables
+the muxer selection to be carried out along with security protocol handshake.
+The proposed solution saves the RTT of multistream selection for muxers.
+
+For more context and the status of this work, please refer to [#426]
+
+
+## Design
+
+### Current connection upgrade process
+
+The current connection upgrade process is described in detail in [connections].
+As shown in this [sequence-chart], after a network connection is established,
+the following will happen to upgrade the conntection to a capable connection.
+
+1. The multistream-selection protocol is run over the connection to select the
+security protocol to be used.
+2. The selected security protocol performs handshaking and establishs a secure
+tunnel
+3. The multistream-selection protocol then will run again for muxer selection.
+4. Connection then is upgraded to a capable connection by the selected muxer.
+
+
+## Improved muxer selection
+
+The security protocol's ability of supporting higher level abstract protocol
+negotiation (for example, TLS's support of ALPN, and Noise's support of Early
+Data) makes it possible to collapse the step 2 and step 3 in the previous
+section into one step. Muxer selection can be performed as part of the security
+protocol handshake, thus there is no need to perform another mutistream
+-selection negotiation for muxer selection.
+
+In order to achieve the above stated goal, each candidate muxer will be
+represented by a protocol name/code, and the candidate muxers are supplied to
+the security protocol's handshake process as a list of protocol names.
+
+If the client and server agree upon the common muxer to be used, then the
+result of the muxer selection is a muxer code represented by the selected
+protocol name. If no agreement is reached upon by the client and server
+then an empty muxer code is returned and the connection upgrade process
+MUST fall back to the multistream-selection protocol to negotiate the muxer.
```

Comment:
1. That makes sense, got rid of the implementation details.
2. The behavior of falling back to multistream-selection instead of failing handshaking is a way of providing backward compatibility and also enhance security. For TLS the backword version carries non-empty ASPN, and for Noise early data is empty. And as the data is carried in variable size extension fields, I prefer to process the data only after the peers are authenticated to enhance security. Instead of introducing more edge cases and optimize for a transient phase, I prefer to simply fallback to multistream-selection in the current situation, if the negotiation is set to fail, it does not affect performace if it fails one RTT later. Open to hear other thoughts.